### DISCUSSION TITLE
Sharing a generated client (private NPM package)

### DISCUSSION THREAD

**Discussion by lukecarr** [prisma/prisma#8936] [Q&A]
Has anyone managed to publish a generated client as an NPM package so multiple projects can use a common schema?

I've taken a look at this [Medium article](https://isidoro-ferreiro.medium.com/share-your-prisma-client-across-projects-44d1c7aca6fd), but at the final step, when I go to import my package into another project, the client is missing all of the typings from my schema.

If anyone can share how they've managed to accomplish what I'm trying to do, it'd be greatly appreciated!

**Comment by sejr** [comment]
I've actually asked about this before (and got nowhere, unfortunately) but would still love to see it supported. +1

**Comment by lynnntropy** [comment]
This isn't quite the same thing as what you're talking about, but on my current project I'm successfully sharing my Prisma client between multiple packages using Yarn workspaces, with `@prisma/client` and `prisma` as workspace-level dependencies. The way I use it is I have a "persistence" package that instantiates a `PrismaClient` and exports it for other packages to import, but I'm sure you could just import the client directly into multiple projects in the workspace, too.

**Reply by nvsd** [reply]
If It's not too much trouble could you provide a repo?

**Reply by GongFu666** [reply]
[Here](https://www.smashingmagazine.com/2019/07/yarn-workspaces-organize-project-codebase-pro/#project-root-workspace) is a nice article which explains the principles how to set it up.

**Comment by juanitoddd** [comment]
Allright, 

I believe this is a valid use case, in which we want to have only one schema File and one Prisma client, but maybe distribute it among microservices. What if those microservices are running inside Docker containers ?. Well, here it goes, its ugly 💩 but functional 🤷🏼‍♂️.

In my arquitecture, I have a docker compose file which has a `prisma` container, in which I have my `schema.prisma` and on which, I run `prisma generate`. So that produces the following folders inside node_modules:

- @prisma folder
- .prisma folder

Having that, then I mount those two folders into my other micro-service container to a generic folder `/prisma` and inside that container:

- `npm install /prisma/@prisma/client `         (Install the package from the mounted @prisma/client folder into the node_modules)
- `ln -s /prisma/.prisma  node_modules/.prisma  `               (symbolinc linking of the mounted .prisma folder into node_modules)

Then, you can call

```
import { PrismaClient } from '@prisma/client'
const prisma = new PrismaClient();

```
And this will reference the one prisma (to rule them all) 😉

**Comment by kaykhan** [comment]
One thing i just tried is you could publish your schema.prisma to an npm package. Then in your other project you could generate a prisma client library from your schema.prisma. This works for me, ive only been using it to access types

**Reply by decoded4620** [reply]
Do you have a sample repo with this setup? We have a mono repo with multiple services which want to share the same prisma client, but we can't seem to deploy, because at runtime, the generated client does NOT look in the correct place for the `schema.prisma` file.  Also, it seems that the file is `loaded` and not a dependency. We've tried suggestions e.g. https://isidoro-ferreiro.medium.com/share-your-prisma-client-across-projects-44d1c7aca6fd but still see the problem during `runtime`. 

```
Error:  no such file or directory, open '/app/schema.prisma'
    at /app/main.js:1:931173
    at /app/main.js:1:944201
    at Object.<anonymous> (/app/main.js:1:944433)
    at Module._compile (node:internal/modules/cjs/loader:1103:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1155:10)
```

**Reply by kaykhan** [reply]
I dont think i have this problem becuase ive only been using it to access types like i mentioned. An altenratie solution for you - might be to introspect prisma in each app within the mono repo - or - write a script which copies schema.prisma from where it currently is to where it should be

**Comment by manuelpoelzl** [comment]
I think that this functionality would be really great.

Currently I have a few controller classes that are used in several of my projects, I would like to merge them into a private NPM package, but currently this is not possible because Prisma cannot be shared, and so I have to adapt the same classes in several of my projects if changes are necessary.

For my part, I would very much appreciate it if this functionality would be available at some point!

**Comment by igolka97** [comment]
I prepared minimal and simple example how I did it. Check out here: https://github.com/igolka97/shared-prisma-client-example

**Reply by DopamineDriven** [reply]
Nicely done

**Reply by jpaylor** [reply]
Thank you. You saved me from jumping out of a window.

**Reply by samoriental** [reply]
ive been fumbling with alternative approaches for a couple hours now. your example is clean and simple. thanks for your work 👍

**Reply by Firgrep** [reply]
Awesome!

**Reply by bamadesigner** [reply]
Thank you! This was super helpful.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
